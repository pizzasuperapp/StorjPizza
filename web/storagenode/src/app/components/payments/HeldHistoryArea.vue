// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <section class="held-history-container">
        <div class="held-history-container__header">
            <p class="held-history-container__header__title">Held Amount History</p>
            <div class="held-history-container__header__selection-area">
                <button
                    name="Select All Stats"
                    class="held-history-container__header__selection-area__item"
                    type="button"
                    :class="{ active: isAllStatsShown }"
                    @click="showAllStats"
                >
                    <p class="held-history-container__header__selection-area__item__label">
                        All Stats
                    </p>
                </button>
                <button
                    name="Select Monthly Breakdown"
                    class="held-history-container__header__selection-area__item"
                    type="button"
                    :class="{ active: !isAllStatsShown }"
                    @click="showMonthlyBreakdown"
                >
                    <p class="held-history-container__header__selection-area__item__label">
                        Monthly Breakdown
                    </p>
                </button>
            </div>
        </div>
        <div class="held-history-container__divider" />
        <HeldHistoryAllStatsTable v-if="isAllStatsShown" />
        <HeldHistoryMonthlyBreakdownTable v-else />
    </section>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import HeldHistoryAllStatsTable from '@/app/components/payments/HeldHistoryAllStatsTable.vue';
import HeldHistoryMonthlyBreakdownTable from '@/app/components/payments/HeldHistoryMonthlyBreakdownTable.vue';

// @vue/component
@Component({
    components: {
        HeldHistoryAllStatsTable,
        HeldHistoryMonthlyBreakdownTable,
    },
})
export default class HeldHistoryArea extends Vue {
    /**
     * Indicates if All Stats state is active.
     */
    public isAllStatsShown = true;

    /**
     * Sets held history table state to All Stats.
     */
    public showAllStats(): void {
        this.isAllStatsShown = true;
    }

    /**
     * Sets held history table state to Monthly Breakdown.
     */
    public showMonthlyBreakdown(): void {
        this.isAllStatsShown = false;
    }
}
</script>

<style scoped lang="scss">
    .held-history-container {
        display: flex;
        flex-direction: column;
        padding: 28px 40px 10px;
        background: var(--block-background-color);
        border: 1px solid var(--block-border-color);
        box-sizing: border-box;
        border-radius: 12px;
        margin: 12px 0 50px;

        &__header {
            display: flex;
            flex-direction: row;
            align-items: flex-start;
            justify-content: space-between;
            height: 40px;

            &__title {
                font-family: 'font_medium', sans-serif;
                font-size: 18px;
                color: var(--regular-text-color);
            }

            &__selection-area {
                display: flex;
                align-items: center;
                justify-content: flex-end;
                height: 100%;

                &__item {
                    display: flex;
                    align-items: flex-start;
                    justify-content: center;
                    cursor: pointer;
                    height: 100%;
                    padding: 0 20px;
                    border-bottom: 3px solid transparent;
                    z-index: 102;

                    &__label {
                        text-align: center;
                        font-size: 16px;
                        color: var(--regular-text-color);
                    }

                    &.active {
                        border-bottom: 3px solid var(--navigation-link-color);

                        &__label {
                            font-size: 16px;
                            color: var(--regular-text-color);
                        }
                    }
                }
            }
        }

        &__divider {
            width: 100%;
            height: 1px;
            background-color: #eaeaea;
        }
    }

    @media screen and (max-width: 870px) {

        .held-history-container {

            &__divider {
                display: none;
            }

            &__header {
                flex-direction: column;
                align-items: flex-start;
                height: auto;

                &__selection-area {
                    width: 100%;
                    height: 41px;
                    margin: 20px 0;

                    &__item {
                        width: calc(50% - 40px);
                        border-bottom: 3px solid #eaeaea;
                    }
                }
            }
        }
    }

    @media screen and (max-width: 600px) {

        .held-history-container {
            padding: 28px 20px 10px;
        }
    }
</style>
