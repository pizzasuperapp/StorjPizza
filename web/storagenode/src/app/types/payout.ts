// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * Holds 'start' and 'end' of payout period range.
 */
export class PayoutInfoRange {
    public constructor(
        public start: PayoutPeriod | null = null,
        public end: PayoutPeriod = new PayoutPeriod(),
    ) {}
}

/**
 * Holds all payout module state.
 */
export class PayoutState {
    public constructor (
        public totalPaystubForPeriod: TotalPaystubForPeriod = new TotalPaystubForPeriod(),
        public periodRange: PayoutInfoRange = new PayoutInfoRange(),
        public totalPayments: TotalPayments = new TotalPayments(),
        public currentMonthEarnings: number = 0,
        public heldPercentage: number = 0,
        public payoutPeriods: PayoutPeriod[] = [],
        public heldHistory: SatelliteHeldHistory[] = [],
        public payoutHistory: SatellitePayoutForPeriod[] = [],
        public payoutHistoryPeriod: string = '',
        public estimation: EstimatedPayout = new EstimatedPayout(),
        public payoutHistoryAvailablePeriods: PayoutPeriod[] = [],
    ) {}
}

export interface StoredMonthsByYear {
    [key: number]: MonthButton[];
}

/**
 * Holds all months names.
 */
export const monthNames = [
    'January', 'February', 'March', 'April',
    'May', 'June', 'July',	'August',
    'September', 'October', 'November',	'December',
];

/**
 * Describes month button entity for calendar.
 */
export class MonthButton {
    public constructor(
        public year: number = 0,
        public index: number = 0,
        public active: boolean = false,
        public selected: boolean = false,
    ) {}

    /**
     * Returns month label depends on index.
     */
    public get name(): string {
        return monthNames[this.index] ? monthNames[this.index].slice(0, 3) : '';
    }
}
