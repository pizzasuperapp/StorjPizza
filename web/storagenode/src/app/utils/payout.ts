// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

// TODO: functions should be moved to related business logic layer
/**
 * Returns held percentage depends on number of months that node is online.
 * @param startedAt date since node is online.
 */
export function getHeldPercentage(startedAt: Date): number {
    const monthsOnline = getMonthsBeforeNow(startedAt);

    switch (true) {
    case monthsOnline < 4:
        return 75;
    case monthsOnline < 7:
        return 50;
    case monthsOnline < 10:
        return 25;
    default:
        return 0;
    }
}

/**
 * Returns number of months passes till now.
 * @param startedAt date since node is online.
 */
export function getMonthsBeforeNow(startedAt: Date): number {
    const now = new Date();
    const yearsDiff =  now.getUTCFullYear() - startedAt.getUTCFullYear();

    return (yearsDiff * 12) + (now.getUTCMonth() - startedAt.getUTCMonth()) + 1;
}
