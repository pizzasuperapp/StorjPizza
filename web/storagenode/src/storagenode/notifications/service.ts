// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { NotificationsApi, NotificationsCursor, NotificationsResponse } from '@/storagenode/notifications/notifications';

/**
 * PayoutService is used to store and handle node paystub information.
 * PayoutService exposes a business logic related to payouts.
 */
export class NotificationsService {
    private readonly api: NotificationsApi;

    public constructor(api: NotificationsApi) {
        this.api = api;
    }

    /**
     * Fetch notifications.
     *
     * @returns notifications response.
     * @throws Error
     */
    public async notifications(index: number, limit?: number): Promise<NotificationsResponse> {
        const cursor = new NotificationsCursor(index, limit);

        return await this.api.get(cursor);
    }

    /**
     * Marks single notification as read on server.
     * @param id
     */
    public async readSingeNotification(id: string): Promise<void> {
        await this.api.read(id);
    }

    /**
     * Marks all notifications as read on server.
     */
    public async readAllNotifications(): Promise<void> {
        await this.api.readAll();
    }
}
