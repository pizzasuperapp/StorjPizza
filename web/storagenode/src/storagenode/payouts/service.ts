// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * PayoutService is used to store and handle node paystub information.
 * PayoutService exposes a business logic related to payouts.
 */
export class PayoutService {
    private readonly payouts: PayoutApi;

    public constructor(api: PayoutApi) {
        this.payouts = api;
    }

    /**
     * Gets summary of paystubs for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async paystubSummaryForPeriod(start: PayoutPeriod | null, end: PayoutPeriod, satelliteId?: string): Promise<TotalPaystubForPeriod> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPaystubForPeriod(paystubs);
    }

    /**
     * Gets held and paid summary for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async totalPayments(start: PayoutPeriod, end: PayoutPeriod, satelliteId: string): Promise<TotalPayments> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPayments(paystubs);
    }

    /**
     * Gets list of payout periods that have paystubs for selected satellite.
     * If satelliteId is not provided returns periods for all satellites.
     * @param satelliteId
     */
    public async availablePeriods(satelliteId: string): Promise<PayoutPeriod[]> {
        return await this.payouts.getPayoutPeriods(satelliteId);
    }

    /**
     * Gets list of payout history items for given period by satellites.
     * @param payoutHistoryPeriod year and month representation
     */
    public async payoutHistory(payoutHistoryPeriod: string): Promise<SatellitePayoutForPeriod[]> {
        return await this.payouts.getPayoutHistory(payoutHistoryPeriod);
    }

    /**
     * Gets list of held history for all satellites.
     */
    public async allSatellitesHeldHistory(): Promise<SatelliteHeldHistory[]> {
        return await this.payouts.getHeldHistory();
    }

    /**
     * Gets estimated payout when no data in paystub.
     * @param satelliteId
     */
    public async estimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        return await this.payouts.getEstimatedPayout(satelliteId);
    }
}
