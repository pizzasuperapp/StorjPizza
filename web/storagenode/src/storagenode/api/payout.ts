// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    PreviousMonthEstimatedPayout,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
} from '@/storagenode/payouts/payouts';
import { HttpClient } from '@/storagenode/utils/httpClient';

/**
 * PayoutHttpApi is a http implementation of Payout API.
 * Exposes all payout-related functionality
 */
export class PayoutHttpApi implements PayoutApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/heldamount';

    /**
     * Fetch paystubs for selected period.
     *
     * @returns paystubs for given period
     * @throws Error
     */
    public async getPaystubsForPeriod(paymentInfoParameters: PaymentInfoParameters): Promise<Paystub[]> {
        let path = `${this.ROOT_PATH}/paystubs/`;

        if (paymentInfoParameters.start) {
            path += paymentInfoParameters.start.period + '/';
        }

        path += paymentInfoParameters.end.period;

        if (paymentInfoParameters.satelliteId) {
            path += '?id=' + paymentInfoParameters.satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held information');
        }

        const responseBody = await response.json() || [];
        const data: any[] = !Array.isArray(responseBody) ? [ responseBody ] : responseBody; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((paystubJson: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new Paystub(
                paystubJson.usageAtRest,
                paystubJson.usageGet,
                paystubJson.usagePut,
                paystubJson.usageGetRepair,
                paystubJson.usagePutRepair,
                paystubJson.usageGetAudit,
                paystubJson.compAtRest,
                paystubJson.compGet,
                paystubJson.compPut,
                paystubJson.compGetRepair,
                paystubJson.compPutRepair,
                paystubJson.compGetAudit,
                paystubJson.surgePercent,
                paystubJson.held,
                paystubJson.owed,
                paystubJson.disposed,
                paystubJson.paid,
                paystubJson.distributed,
            );
        });
    }

    /**
     * Fetches available payout periods.
     *
     * @returns payout periods list
     * @throws Error
     */
    public async getPayoutPeriods(id: string): Promise<PayoutPeriod[]> {
        let path = `${this.ROOT_PATH}/periods`;

        if (id) {
            path += '?id=' + id;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout periods');
        }

        const result = await response.json() || [];

        return result.map(period => {
            return PayoutPeriod.fromString(period);
        });
    }

    /**
     * Fetch payout history for given period.
     *
     * @returns payout information
     * @throws Error
     */
    public async getPayoutHistory(period: string): Promise<SatellitePayoutForPeriod[]> {
        const path = `${this.ROOT_PATH}/payout-history/${period}`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((payoutHistoryItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatellitePayoutForPeriod(
                payoutHistoryItem.satelliteID,
                payoutHistoryItem.satelliteURL,
                payoutHistoryItem.age,
                payoutHistoryItem.earned,
                payoutHistoryItem.surge,
                payoutHistoryItem.surgePercent,
                payoutHistoryItem.held,
                payoutHistoryItem.afterHeld,
                payoutHistoryItem.disposed,
                payoutHistoryItem.paid,
                payoutHistoryItem.receipt,
                payoutHistoryItem.isExitComplete,
                payoutHistoryItem.heldPercent,
                payoutHistoryItem.distributed,
            );
        });
    }

    /**
     * Fetch total payout information.
     *
     * @returns total payout information
     * @throws Error
     */
    public async getHeldHistory(): Promise<SatelliteHeldHistory[]> {
        const path = `${this.ROOT_PATH}/held-history`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((historyItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatelliteHeldHistory(
                historyItem.satelliteID,
                historyItem.satelliteName,
                historyItem.holdForFirstPeriod,
                historyItem.holdForSecondPeriod,
                historyItem.holdForThirdPeriod,
                historyItem.totalHeld,
                historyItem.totalDisposed,
                new Date(historyItem.joinedAt),
            );
        });
    }

    /**
     * Fetch estimated payout information.
     *
     * @returns estimated payout information
     * @throws Error
     */
    public async getEstimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        let path = '/api/sno/estimated-payout';

        if (satelliteId) {
            path += '?id=' + satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get estimated payout information');
        }

        const data: any = await response.json() || new EstimatedPayout(); // eslint-disable-line @typescript-eslint/no-explicit-any

        return new EstimatedPayout(
            new PreviousMonthEstimatedPayout(
                data.currentMonth.egressBandwidth,
                data.currentMonth.egressBandwidthPayout,
                data.currentMonth.egressRepairAudit,
                data.currentMonth.egressRepairAuditPayout,
                data.currentMonth.diskSpace,
                data.currentMonth.diskSpacePayout,
                data.currentMonth.heldRate,
                data.currentMonth.payout,
                data.currentMonth.held,
            ),
            new PreviousMonthEstimatedPayout(
                data.previousMonth.egressBandwidth,
                data.previousMonth.egressBandwidthPayout,
                data.previousMonth.egressRepairAudit,
                data.previousMonth.egressRepairAuditPayout,
                data.previousMonth.diskSpace,
                data.previousMonth.diskSpacePayout,
                data.previousMonth.heldRate,
                data.previousMonth.payout,
                data.previousMonth.held,
            ),
            data.currentMonthExpectations,
        );
    }
}
