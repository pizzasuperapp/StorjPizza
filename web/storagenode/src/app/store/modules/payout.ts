// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/app/store';
import {
    PayoutInfoRange,
    PayoutState,
} from '@/app/types/payout';
import { StorageNodeState } from '@/app/types/sno';
import { getHeldPercentage } from '@/app/utils/payout';
import { SizeBreakpoints } from '@/private/memory/size';
import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';
import { PayoutService } from '@/storagenode/payouts/service';

export const PAYOUT_MUTATIONS = {
    SET_PAYOUT_INFO: 'SET_PAYOUT_INFO',
    SET_RANGE: 'SET_RANGE',
    SET_TOTAL: 'SET_TOTAL',
    SET_HELD_PERCENT: 'SET_HELD_PERCENT',
    SET_HELD_HISTORY: 'SET_HELD_HISTORY',
    SET_ESTIMATION: 'SET_ESTIMATION',
    SET_PERIODS: 'SET_PERIODS',
    SET_PAYOUT_HISTORY: 'SET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
    SET_PAYOUT_HISTORY_AVAILABLE_PERIODS: 'SET_PAYOUT_HISTORY_AVAILABLE_PERIODS',
};

export const PAYOUT_ACTIONS = {
    GET_PAYOUT_INFO: 'GET_PAYOUT_INFO',
    SET_PERIODS_RANGE: 'SET_PERIODS_RANGE',
    GET_TOTAL: 'GET_TOTAL',
    GET_HELD_HISTORY: 'GET_HELD_HISTORY',
    GET_ESTIMATION: 'GET_ESTIMATION',
    GET_PERIODS: 'GET_PERIODS',
    GET_PAYOUT_HISTORY: 'GET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
};

// TODO: move to config in storagenode/payouts
export const BANDWIDTH_DOWNLOAD_PRICE_PER_TB = 2000;
export const BANDWIDTH_REPAIR_PRICE_PER_TB = 1000;
export const DISK_SPACE_PRICE_PER_TB = 150;

interface PayoutContext {
    rootState: {
        node: StorageNodeState;
    };
    state: PayoutState;
    commit: (string, ...unknown) => void;
}

/**
 * creates notifications module with all dependencies
 *
 * @param service - payments service
 */
export function newPayoutModule(service: PayoutService): StoreModule<PayoutState> {
    return {
        state: new PayoutState(),
        mutations: {
            [PAYOUT_MUTATIONS.SET_PAYOUT_INFO](state: PayoutState, totalPaystubForPeriod: TotalPaystubForPeriod): void {
                state.totalPaystubForPeriod = totalPaystubForPeriod;
            },
            [PAYOUT_MUTATIONS.SET_TOTAL](state: PayoutState, totalPayments: TotalPayments): void {
                state.totalPayments = totalPayments;
                state.currentMonthEarnings = totalPayments.currentMonthEarnings;
            },
            [PAYOUT_MUTATIONS.SET_RANGE](state: PayoutState, periodRange: PayoutInfoRange): void {
                state.periodRange = periodRange;
            },
            [PAYOUT_MUTATIONS.SET_HELD_PERCENT](state: PayoutState, heldPercentage: number): void {
                state.heldPercentage = heldPercentage;
            },
            [PAYOUT_MUTATIONS.SET_HELD_HISTORY](state: PayoutState, heldHistory: SatelliteHeldHistory[]): void {
                state.heldHistory = heldHistory;
            },
            [PAYOUT_MUTATIONS.SET_ESTIMATION](state: PayoutState, estimatedInfo: EstimatedPayout): void {
                state.estimation = estimatedInfo;
            },
            [PAYOUT_MUTATIONS.SET_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutPeriods = periods;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY](state: PayoutState, payoutHistory: SatellitePayoutForPeriod[]): void {
                state.payoutHistory = payoutHistory;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD](state: PayoutState, period: string): void {
                state.payoutHistoryPeriod = period;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutHistoryAvailablePeriods = periods;
            },
        },
        actions: {
            [PAYOUT_ACTIONS.GET_PAYOUT_INFO]: async function ({ commit, state, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const totalPaystubForPeriod = await service.paystubSummaryForPeriod(
                    state.periodRange.start,
                    state.periodRange.end,
                    satelliteId,
                );

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_INFO, totalPaystubForPeriod);
            },
            [PAYOUT_ACTIONS.GET_TOTAL]: async function ({ commit, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const now = new Date();
                const start = new PayoutPeriod(rootState.node.selectedSatellite.joinDate.getUTCFullYear(), rootState.node.selectedSatellite.joinDate.getUTCMonth());
                const end = new PayoutPeriod(now.getUTCFullYear(), now.getUTCMonth());

                const totalPayments = await service.totalPayments(start, end, satelliteId);

                // TODO: move to service
                const currentBandwidthDownload = (rootState.node.egressChartData || [])
                    .map(data => data.egress.usage)
                    .reduce((previous, current) => previous + current, 0);

                const currentBandwidthAuditAndRepair = (rootState.node.egressChartData || [])
                    .map(data => data.egress.audit + data.egress.repair)
                    .reduce((previous, current) => previous + current, 0);

                const approxHourInMonth = 720;
                const currentDiskSpace = (rootState.node.storageChartData || [])
                    .map(data => data.atRestTotal)
                    .reduce((previous, current) => previous + current, 0) / approxHourInMonth;

                const thisMonthEarnings = (currentBandwidthDownload * BANDWIDTH_DOWNLOAD_PRICE_PER_TB
                    + currentBandwidthAuditAndRepair * BANDWIDTH_REPAIR_PRICE_PER_TB
                    + currentDiskSpace * DISK_SPACE_PRICE_PER_TB) / SizeBreakpoints.TB;

                totalPayments.setCurrentMonthEarnings(thisMonthEarnings);

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_TOTAL, totalPayments);
            },
            [PAYOUT_ACTIONS.SET_PERIODS_RANGE]: function ({ commit }: PayoutContext, periodRange: PayoutInfoRange): void {
                commit(PAYOUT_MUTATIONS.SET_RANGE, periodRange);
            },
            [PAYOUT_ACTIONS.GET_HELD_HISTORY]: async function ({ commit }: PayoutContext): Promise<void> {
                const heldHistory = await service.allSatellitesHeldHistory();

                commit(PAYOUT_MUTATIONS.SET_HELD_HISTORY, heldHistory);
            },
            [PAYOUT_ACTIONS.GET_PERIODS]: async function ({commit}: PayoutContext, satelliteId = ''): Promise<void> {
                const periods = await service.availablePeriods(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_PERIODS, periods);

                if (!satelliteId) {
                    commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS, periods);
                }
            },
            [PAYOUT_ACTIONS.GET_ESTIMATION]: async function ({ commit }: PayoutContext, satelliteId = ''): Promise<void> {
                const estimatedInfo = await service.estimatedPayout(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_ESTIMATION, estimatedInfo);
            },
            [PAYOUT_ACTIONS.GET_PAYOUT_HISTORY]: async function ({ commit, state }: PayoutContext): Promise<void> {
                if (!state.payoutHistoryPeriod) return;

                const payoutHistory = await service.payoutHistory(state.payoutHistoryPeriod);

                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY, payoutHistory);
            },
            [PAYOUT_ACTIONS.SET_PAYOUT_HISTORY_PERIOD]: function ({ commit }: PayoutContext, period: string): void {
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD, period);
            },
        },
        getters: {
            totalPaidForPayoutHistoryPeriod: (state: PayoutState): number => {
                return state.payoutHistory.map(data => data.paid)
                    .reduce((previous, current) => previous + current, 0);
            },
        },
    };
}
