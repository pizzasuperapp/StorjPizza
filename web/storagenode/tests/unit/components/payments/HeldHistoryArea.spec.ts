// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import HeldHistoryArea from '@/app/components/payments/HeldHistoryArea.vue';

import { newPayoutModule } from '@/app/store/modules/payout';
import { PayoutHttpApi } from '@/storagenode/api/payout';
import { PayoutService } from '@/storagenode/payouts/service';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

const payoutApi = new PayoutHttpApi();
const payoutService = new PayoutService(payoutApi);
const payoutModule = newPayoutModule(payoutService);

const store = new Vuex.Store({ modules: { payoutModule }});

describe('HeldHistoryArea', (): void => {
    it('renders correctly',  (): void => {
        const wrapper = shallowMount(HeldHistoryArea, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('changes state correctly',  async (): Promise<void> => {
        const wrapper = shallowMount(HeldHistoryArea, {
            store,
            localVue,
        });

        await wrapper.findAll('.held-history-container__header__selection-area__item').at(1).trigger('click');

        expect(wrapper).toMatchSnapshot();

        await wrapper.findAll('.held-history-container__header__selection-area__item').at(0).trigger('click');

        expect(wrapper).toMatchSnapshot();
    });
});
