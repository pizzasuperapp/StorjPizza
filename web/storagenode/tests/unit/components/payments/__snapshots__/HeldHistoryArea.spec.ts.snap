// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeldHistoryArea changes state correctly 1`] = `
<section class="held-history-container">
  <div class="held-history-container__header">
    <p class="held-history-container__header__title">Held Amount History</p>
    <div class="held-history-container__header__selection-area"><button name="Select All Stats" type="button" class="held-history-container__header__selection-area__item">
        <p class="held-history-container__header__selection-area__item__label">
          All Stats
        </p>
      </button> <button name="Select Monthly Breakdown" type="button" class="held-history-container__header__selection-area__item active">
        <p class="held-history-container__header__selection-area__item__label">
          Monthly Breakdown
        </p>
      </button></div>
  </div>
  <div class="held-history-container__divider"></div>
  <heldhistorymonthlybreakdowntable-stub></heldhistorymonthlybreakdowntable-stub>
</section>
`;

exports[`HeldHistoryArea changes state correctly 2`] = `
<section class="held-history-container">
  <div class="held-history-container__header">
    <p class="held-history-container__header__title">Held Amount History</p>
    <div class="held-history-container__header__selection-area"><button name="Select All Stats" type="button" class="held-history-container__header__selection-area__item active">
        <p class="held-history-container__header__selection-area__item__label">
          All Stats
        </p>
      </button> <button name="Select Monthly Breakdown" type="button" class="held-history-container__header__selection-area__item">
        <p class="held-history-container__header__selection-area__item__label">
          Monthly Breakdown
        </p>
      </button></div>
  </div>
  <div class="held-history-container__divider"></div>
  <heldhistoryallstatstable-stub></heldhistoryallstatstable-stub>
</section>
`;

exports[`HeldHistoryArea renders correctly 1`] = `
<section class="held-history-container">
  <div class="held-history-container__header">
    <p class="held-history-container__header__title">Held Amount History</p>
    <div class="held-history-container__header__selection-area"><button name="Select All Stats" type="button" class="held-history-container__header__selection-area__item active">
        <p class="held-history-container__header__selection-area__item__label">
          All Stats
        </p>
      </button> <button name="Select Monthly Breakdown" type="button" class="held-history-container__header__selection-area__item">
        <p class="held-history-container__header__selection-area__item__label">
          Monthly Breakdown
        </p>
      </button></div>
  </div>
  <div class="held-history-container__divider"></div>
  <heldhistoryallstatstable-stub></heldhistoryallstatstable-stub>
</section>
`;
